### DISCUSSION TITLE
Proposal: Color Support for Drop Shadows

### DISCUSSION THREAD

**Discussion by surjithctly** [tailwindlabs/tailwindcss#4744] [Ideas]
It would be a great feature if **tailwindcss** supports colors for drop-shadows. 

Currently, it uses a default gray color without an option to change color or opacity. 

Since the launch of JIT, I think it's now easier to add these features to the code. 
I would prefer something like this.

```html
<div class="shadow-xl shadow-indigo-300">
...
<div class="shadow-xl shadow-gray-300/50">
...
<div class="shadow-xl/50 shadow-indigo-300">
``` 

It could even be expanded for multiple drop-shadows. 

```html
<div class="shadow-md shadow-md-gray-300/50 shadow-xl shadow-xl-gray-100/20 ">
```

And JIT

```html
<div class="shadow-[5px,5px,25px] shadow-indigo-300">
...
<div class="shadow-[5px,5px,25px,#aabbcc]">
```

If you like this proposal please do a 👍

**Comment by VicGUTT** [comment]
I like how you think !

I took a small stab at this a couple weeks back and came up with a similar API although smaller in scope.

```html
<!-- What I came up with -->
<button class="shadow-2xl shadow-blue-600">
<button class="drop-shadow-2xl drop-shadow-blue-600">
```
Having something like this:

```html
<button class="shadow-2xl-blue-600">
```

was easier to do but not very "Tailwind" like in my opinion and also not very composable.

The way I went about it was to generate _colored versions_ of what was defined in the user's theme.
This brought up a handful of challenges, and so I had to come up with helper functions like:

```js
parseShadowValue('inset 0 20px 25px -5px rgba(0, 0, 0, 0.1)'); // imagine the shadow value coming from the user's theme

// output
[{
    offsetX: '0',
    offsetY: '20px',
    blurRadius: '25px',
    spreadRadius: '-5px',
    inset: true,
    color: new TinyColor('rgba(0, 0, 0, 0.1)'),
}];

// -----------------------------------------------------------------

replaceShadowColorWithCssVar({ sm: '0 1px 2px 0 rgba(0, 0, 0, 0.05)' }, 'box');

// output
{
    sm: ['0 1px 2px 0 rgba(var(--twcs-box-shadow-color, 0, 0, 0), 0.05)'],
}
``` 

The goal being to extract the provided values and replace the colors with custom CSS variables so that `shadow-2xl` would work as normal but `shadow-blue-600` would simply define a different value for the `--twcs-box-shadow-color` CSS variable.

Another challenge was making the resulting shadows look good. As the values defined for "dark/regular" shadows does not provide a 1/1 design awesomeness with "colored shadows", so I made some small opinionated choices.

Take a look at the discussion [#4735](https://github.com/tailwindlabs/tailwindcss/discussions/4735) published this morning for updates.
I happily welcome brain powers on the matter :)

**Reply by surjithctly** [reply]
That's nice approach. Would have been great if its composable. Also I think Tailwind CSS should implement this in the core since it has been a problem on every project.

**Comment by VicGUTT** [comment]
Sure, I'd also like it to be in core, but we can start getting the ball rolling :)

Also, like I mentioned, simply changing the provided shadows' colors doesn't always yield good results visually.
So: `0 25px 50px -12px rgba(0, 0, 0, 0.25)` _(default .shadow-2xl)_ looks good, but: `0 25px 50px -12px rgba(x, x, x, 0.25)` not necessarily. We'd need to play with the values a bit to get to a satisfying result. Therein lies the challenge in my opinion.

I agree, it'd be cool to have the shadows as composable as possible.
I was thinking providing utilities like:
- **offsetX**: `.shadow-x-2xs`, `.shadow-x-20`, `.shadow-x-[77px]`, ...
- **offsetY**: `.shadow-y-lg`, `.shadow-y-8`, `.shadow-y-[9rem]`, ...
- **blurRadius**: `.shadow-blur-md`, `.shadow-blur-12`, `.shadow-blur-[25px]`, ...
- **spreadRadius**: `.shadow-spread-md`, `.shadow-spread-12`, `.shadow-spread-[-5px]`, ...
- **inset**: `.shadow-inset`
- **color**: `.shadow-blue-500`, `.shadow-[#fff]`
